### PR TITLE
[HOPSWORKS-2280] Save statistics configuration also for training datasets

### PR DESCRIPTION
**Pull request by moritzmeister**
There is one issue @SirOibaf the following columns should have been removed in the 2.0 release. Probably a bad refactor. they don't hurt in the table, however, I add drop column statements for them in the 2.1.0 update sql, which will fail if they exist. I am wondering if we should just leave them for 2.0 clusters or fix the release branch etc?

```
  `cluster_analysis_enabled` TINYINT(1) NOT NULL DEFAULT '1',
  `num_clusters` int(11) NOT NULL DEFAULT '5',
  `num_bins` INT(11) NOT NULL DEFAULT '20',
  `corr_method` VARCHAR(50) NOT NULL DEFAULT 'pearson',
```

### CODE REVIEW

**Comment by SirOibaf** [comment]
> There is one issue @SirOibaf the following columns should have been removed in the 2.0 release. Probably a bad refactor. they don't hurt in the table, however, I add drop column statements for them in the 2.1.0 update sql, which will fail if they exist. I am wondering if we should just leave them for 2.0 clusters or fix the release branch etc?
> 
> ```
>   `cluster_analysis_enabled` TINYINT(1) NOT NULL DEFAULT '1',
>   `num_clusters` int(11) NOT NULL DEFAULT '5',
>   `num_bins` INT(11) NOT NULL DEFAULT '20',
>   `corr_method` VARCHAR(50) NOT NULL DEFAULT 'pearson',
> ```

No it's not safe to fix the release branch, but in the update file you can do `DROP COLUMN IF EXISTS` so that the migration doesn't fail if the column is missing already

File: files/default/sql/ddl/2.1.0__initial_tables.sql [hunk 1]
```diff
@@ -1516,6 +1530,9 @@ CREATE TABLE `training_dataset` (
   `training_dataset_type`   INT(11) NOT NULL DEFAULT '0',
   `seed` BIGINT(11) NULL,
   `query` TINYINT(1) NOT NULL DEFAULT '0',
+  `desc_stats_enabled` TINYINT(1) NOT NULL DEFAULT '1',
+  `feat_corr_enabled` TINYINT(1) NOT NULL DEFAULT '1',
+  `feat_hist_enabled` TINYINT(1) NOT NULL DEFAULT '1',
```
Comment (SirOibaf) [inline files/default/sql/ddl/2.1.0__initial_tables.sql:1535]: Why are these here?

File: files/default/sql/ddl/updates/2.1.0.sql [hunk 2]
```diff
@@ -147,4 +147,65 @@ CREATE TABLE `on_demand_option` (
...
+       `desc_stats_enabled` as `descriptive`,
+       `feat_corr_enabled` as `correlations`,
+       `feat_hist_enabled` as `histograms`
+    FROM `hopsworks`.`feature_group`;
+
+ALTER TABLE `hopsworks`.`feature_group`
+    DROP COLUMN `cluster_analysis_enabled`,
+    DROP COLUMN `num_clusters`,
+    DROP COLUMN `num_bins`,
+    DROP COLUMN `corr_method`,
```
Comment (SirOibaf) [inline files/default/sql/ddl/updates/2.1.0.sql:177]: `IF EXISTS`

File: files/default/sql/ddl/updates/2.1.0.sql [hunk 3]
```diff
@@ -147,4 +147,65 @@ CREATE TABLE `on_demand_option` (
...
+  `name` varchar(500) COLLATE latin1_general_cs DEFAULT NULL,
+  PRIMARY KEY (`id`),
+  KEY `statistics_config_id` (`statistics_config_id`),
+  CONSTRAINT `statistics_config_fk` FOREIGN KEY (`statistics_config_id`) REFERENCES `statistics_config` (`id`) ON DELETE CASCADE ON UPDATE NO ACTION
+) ENGINE=ndbcluster DEFAULT CHARSET=latin1 COLLATE=latin1_general_cs;
+
+INSERT INTO `hopsworks`.`statistic_columns` (`statistics_config_id`, `name`)
+SELECT * FROM `hopsworks`.`tmp_stats_columns`;
+
+DROP TABLE `hopsworks`.`tmp_stats_columns`;
```
Comment (SirOibaf) [inline files/default/sql/ddl/updates/2.1.0.sql:210]: This looks complicated. Here's what I'd do:
- add `statistics_config_id` to the `statistics_column` table.
- Update the rows so that `statistics_config_id` matches the correct value based on the `feature_group_id` already present
- Drop the `feature_group_id` column.

File: files/default/sql/ddl/updates/undo/2.1.0__undo.sql [hunk 4]
```diff
@@ -112,3 +112,39 @@ ALTER TABLE `on_demand_feature_group`
...
+  `name` varchar(500) COLLATE latin1_general_cs DEFAULT NULL,
+  PRIMARY KEY (`id`),
+  KEY `feature_group_id` (`feature_group_id`),
+  CONSTRAINT `statistic_column_fk` FOREIGN KEY (`feature_group_id`) REFERENCES `feature_group` (`id`) ON DELETE CASCADE ON UPDATE NO ACTION
+) ENGINE=ndbcluster DEFAULT CHARSET=latin1 COLLATE=latin1_general_cs;
+
+INSERT INTO `hopsworks`.`statistic_columns` (`feature_group_id`, `name`)
+SELECT * FROM `hopsworks`.`tmp_stats_columns`;
+
+DROP TABLE `hopsworks`.`tmp_stats_columns`;
```
Comment (SirOibaf) [inline files/default/sql/ddl/updates/undo/2.1.0__undo.sql:150]: Same comment as the above.

**Review by SirOibaf — CHANGES_REQUESTED** [review]

**Comment by SirOibaf** [comment]
> > There is one issue @SirOibaf the following columns should have been removed in the 2.0 release. Probably a bad refactor. they don't hurt in the table, however, I add drop column statements for them in the 2.1.0 update sql, which will fail if they exist. I am wondering if we should just leave them for 2.0 clusters or fix the release branch etc?
> > ```
> >   `cluster_analysis_enabled` TINYINT(1) NOT NULL DEFAULT '1',
> >   `num_clusters` int(11) NOT NULL DEFAULT '5',
> >   `num_bins` INT(11) NOT NULL DEFAULT '20',
> >   `corr_method` VARCHAR(50) NOT NULL DEFAULT 'pearson',
> > ```
> 
> No it's not safe to fix the release branch, but in the update file you can do `DROP COLUMN IF EXISTS` so that the migration doesn't fail if the column is missing already

Actually MySQL doesn't seem to support the `DROP COLUMN IF EXISTS` statement - https://chase-seibert.github.io/blog/2010/01/15/mysql-drop-column-if-exists.html - we need to query the information schema.

File: files/default/sql/ddl/updates/2.1.0.sql (hunk 2 above)
Comment (moritzmeister) [reply files/default/sql/ddl/updates/2.1.0.sql:177]: yeah mysql doesn't support that, will do it with querying the schema

File: files/default/sql/ddl/2.1.0__initial_tables.sql (hunk 1 above)
Comment (moritzmeister) [reply files/default/sql/ddl/2.1.0__initial_tables.sql]: thanks, that's a leftover, removed them.

File: files/default/sql/ddl/updates/2.1.0.sql (hunk 3 above)
Comment (moritzmeister) [reply files/default/sql/ddl/updates/2.1.0.sql]: simplified it as suggested.

**Review by SirOibaf — APPROVED** [review]
